[ICE]: `item_ident: no name for DefPath`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_middle/src/ty/mod.rs:1692:13: item_ident: no name for DefPath { data: [DisambiguatedDefPathData { data: Impl, disambiguator: 0 }, DisambiguatedDefPathData { data: AnonAssocTy("bar"), disambiguator: 0 }], krate: crate0 }', 'error: internal compiler error: compiler/rustc_middle/src/ty/mod.rs:1692:13: item_ident: no name for DefPath { data: [DisambiguatedDefPathData { data: Impl, disambiguator: 0 }, DisambiguatedDefPathData { data: AnonAssocTy("bar"), disambiguator: 0 }], krate: crate0 }'
File: /tmp/IM/2/a.rs
-->

auto-reduced (treereduce-rust):
````rust
trait Item {
    final fn bar() -> impl std::clone::Clone {
        todo!()
    }
}
struct Foo;
impl Item for Foo {
    fn bar() -> impl std::clone::Clone {}
}

````

original:
````rust
#![feature(final_associated_functions)]
trait Item {
    final fn bar() -> impl std::clone::Clone {
        todo!()
    }
}
struct Foo;
impl Item for Foo {
    fn bar() -> impl std::clone::Clone {}
}

fn main() {}

````

Version information
````
rustc 1.98.0-nightly (4eb62535f 2026-07-05)
binary: rustc
commit-hash: 4eb62535fc12dd1a63fd43a4173e224e79313c4d
commit-date: 2026-07-05
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/4eb62535fc12dd1a63fd43a4173e224e79313c4d/compiler/rustc_middle/src/ty/mod.rs#L1686-L1698

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: `final` on trait functions is experimental
 --> /tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs:2:5
  |
2 |     final fn bar() -> impl std::clone::Clone {
  |     ^^^^^
  |
  = note: see issue #131179 <https://github.com/rust-lang/rust/issues/131179> for more information
  = help: add `#![feature(final_associated_functions)]` to the crate attributes to enable
  = note: this compiler was built on 2026-07-05; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs:9:2
  |
9 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs`

error: cannot override `bar` because it already has a `final` definition in the trait
 --> /tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs:8:5
  |
8 |     fn bar() -> impl std::clone::Clone {}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: `bar` is marked final here
 --> /tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs:2:5
  |
2 |     final fn bar() -> impl std::clone::Clone {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: /rustc-dev/4eb62535fc12dd1a63fd43a4173e224e79313c4d/compiler/rustc_middle/src/ty/mod.rs:1692:13: item_ident: no name for DefPath { data: [DisambiguatedDefPathData { data: Impl, disambiguator: 0 }, DisambiguatedDefPathData { data: AnonAssocTy("bar"), disambiguator: 0 }], krate: crate0 }


thread 'rustc' (3120148) panicked at /rustc-dev/4eb62535fc12dd1a63fd43a4173e224e79313c4d/compiler/rustc_middle/src/ty/mod.rs:1692:13:
Box<dyn Any>
stack backtrace:
   0:     0x7f3b410df7d6 - <<std[17eae023b1d79ed6]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[458a8607a3cebf18]::fmt::Display>::fmt
   1:     0x7f3b4180ed08 - core[458a8607a3cebf18]::fmt::write
   2:     0x7f3b410f4bbc - <std[17eae023b1d79ed6]::sys::stdio::unix::Stderr as std[17eae023b1d79ed6]::io::Write>::write_fmt
   3:     0x7f3b410b3c3a - std[17eae023b1d79ed6]::panicking::default_hook::{closure#0}
   4:     0x7f3b410d1f23 - std[17eae023b1d79ed6]::panicking::default_hook
   5:     0x7f3b400c2ec8 - std[17eae023b1d79ed6]::panicking::update_hook::<alloc[73ddf1fa857a38ac]::boxed::Box<rustc_driver_impl[db55e0879ee03706]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f3b410d23c2 - std[17eae023b1d79ed6]::panicking::panic_with_hook
   7:     0x7f3b400f09f1 - std[17eae023b1d79ed6]::panicking::begin_panic::<rustc_errors[5475147e90673a3a]::ExplicitBug>::{closure#0}
   8:     0x7f3b400e9766 - std[17eae023b1d79ed6]::sys::backtrace::__rust_end_short_backtrace::<std[17eae023b1d79ed6]::panicking::begin_panic<rustc_errors[5475147e90673a3a]::ExplicitBug>::{closure#0}, !>
   9:     0x7f3b400e77a7 - std[17eae023b1d79ed6]::panicking::begin_panic::<rustc_errors[5475147e90673a3a]::ExplicitBug>
  10:     0x7f3b400fbb81 - <rustc_errors[5475147e90673a3a]::diagnostic::BugAbort as rustc_errors[5475147e90673a3a]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x7f3b40675df9 - rustc_middle[7033358a29c6e3a]::util::bug::opt_span_bug_fmt::<rustc_span[59bd5a5e00db998f]::span_encoding::Span>::{closure#0}
  12:     0x7f3b40675f62 - rustc_middle[7033358a29c6e3a]::ty::context::tls::with_opt::<rustc_middle[7033358a29c6e3a]::util::bug::opt_span_bug_fmt<rustc_span[59bd5a5e00db998f]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  13:     0x7f3b40661c1b - rustc_middle[7033358a29c6e3a]::ty::context::tls::with_context_opt::<rustc_middle[7033358a29c6e3a]::ty::context::tls::with_opt<rustc_middle[7033358a29c6e3a]::util::bug::opt_span_bug_fmt<rustc_span[59bd5a5e00db998f]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  14:     0x7f3b3dc935c4 - rustc_middle[7033358a29c6e3a]::util::bug::bug_fmt
  15:     0x7f3b3eb4c22b - <rustc_middle[7033358a29c6e3a]::ty::context::TyCtxt>::item_ident::<rustc_span[59bd5a5e00db998f]::def_id::DefId>
  16:     0x7f3b426facbb - rustc_hir_analysis[65818f7540dc9cfd]::check::check::check_item_type
  17:     0x7f3b426ea770 - rustc_hir_analysis[65818f7540dc9cfd]::check::wfcheck::check_well_formed
  18:     0x7f3b426ea751 - rustc_query_impl[243cebd7d7019491]::query_impl::check_well_formed::invoke_provider_fn::__rust_begin_short_backtrace
  19:     0x7f3b426e9d31 - rustc_query_impl[243cebd7d7019491]::execution::try_execute_query::<rustc_data_structures[fce1113a2776e9de]::vec_cache::VecCache<rustc_span[59bd5a5e00db998f]::def_id::LocalDefId, rustc_middle[7033358a29c6e3a]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[7033358a29c6e3a]::dep_graph::graph::DepNodeIndex>, false>
  20:     0x7f3b426e9ac7 - rustc_query_impl[243cebd7d7019491]::query_impl::check_well_formed::execute_query_non_incr::__rust_end_short_backtrace
  21:     0x7f3b426e6545 - rustc_hir_analysis[65818f7540dc9cfd]::check::wfcheck::check_type_wf
  22:     0x7f3b426e6453 - rustc_query_impl[243cebd7d7019491]::query_impl::check_type_wf::invoke_provider_fn::__rust_begin_short_backtrace
  23:     0x7f3b4292e0c6 - rustc_query_impl[243cebd7d7019491]::execution::try_execute_query::<rustc_middle[7033358a29c6e3a]::query::caches::SingleCache<rustc_middle[7033358a29c6e3a]::query::erase::ErasedData<[u8; 1usize]>>, false>
  24:     0x7f3b4292dea9 - rustc_query_impl[243cebd7d7019491]::query_impl::check_type_wf::execute_query_non_incr::__rust_end_short_backtrace
  25:     0x7f3b41c1721a - rustc_hir_analysis[65818f7540dc9cfd]::check_crate
  26:     0x7f3b41c1b83a - rustc_interface[7d9099d0bfcaa9c4]::passes::analysis
  27:     0x7f3b4292c647 - rustc_query_impl[243cebd7d7019491]::execution::try_execute_query::<rustc_middle[7033358a29c6e3a]::query::caches::SingleCache<rustc_middle[7033358a29c6e3a]::query::erase::ErasedData<[u8; 0usize]>>, false>
  28:     0x7f3b4292c2a9 - rustc_query_impl[243cebd7d7019491]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  29:     0x7f3b428fffc0 - rustc_interface[7d9099d0bfcaa9c4]::interface::run_compiler::<(), rustc_driver_impl[db55e0879ee03706]::run_compiler::{closure#0}>::{closure#1}
  30:     0x7f3b4293cc3a - std[17eae023b1d79ed6]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[7d9099d0bfcaa9c4]::util::run_in_thread_with_globals<rustc_interface[7d9099d0bfcaa9c4]::util::run_in_thread_pool_with_globals<rustc_interface[7d9099d0bfcaa9c4]::interface::run_compiler<(), rustc_driver_impl[db55e0879ee03706]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  31:     0x7f3b4293c9ed - <std[17eae023b1d79ed6]::thread::lifecycle::spawn_unchecked<rustc_interface[7d9099d0bfcaa9c4]::util::run_in_thread_with_globals<rustc_interface[7d9099d0bfcaa9c4]::util::run_in_thread_pool_with_globals<rustc_interface[7d9099d0bfcaa9c4]::interface::run_compiler<(), rustc_driver_impl[db55e0879ee03706]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[458a8607a3cebf18]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  32:     0x7f3b42940e10 - <std[17eae023b1d79ed6]::sys::thread::unix::Thread>::new::thread_start
  33:     0x7f3b3c2981b9 - <unknown>
  34:     0x7f3b3c31d21c - <unknown>
  35:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.98.0-nightly (4eb62535f 2026-07-05) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs:7:1: 7:18>` is well-formed
#1 [check_type_wf] checking that types are well-formed
#2 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.VS8FQIl5X6bs/rustc_testrunner_tmpdir_reporting.2gLCRH8QPhGz/mvce.rs:7:1: 7:18>` is well-formed
#1 [check_type_wf] checking that types are well-formed
#2 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-final_associated_functions

--- a/compiler/rustc_hir_analysis/src/check/check.rs
+++ b/compiler/rustc_hir_analysis/src/check/check.rs
@@ -1254,7 +1254,7 @@ fn check_overriding_final_trait_item<'tcx>(
     trait_item: ty::AssocItem,
     impl_item: ty::AssocItem,
 ) {
-    if trait_item.defaultness(tcx).is_final() {
+    if trait_item.is_fn() && trait_item.defaultness(tcx).is_final() {
         tcx.dcx().emit_err(diagnostics::OverridingFinalTraitFunction {
             impl_span: tcx.def_span(impl_item.def_id),
             trait_span: tcx.def_span(trait_item.def_id),

--- a/tests/ui/traits/final/overriding-rpitit.rs
+++ b/tests/ui/traits/final/overriding-rpitit.rs
@@ -1,0 +1,18 @@
+#![feature(final_associated_functions)]
+
+// Regression test for https://github.com/rust-lang/rust/issues/158824.
+
+trait Item {
+    final fn bar() -> impl Clone {
+        todo!()
+    }
+}
+
+struct Foo;
+
+impl Item for Foo {
+    fn bar() -> impl Clone {}
+    //~^ ERROR cannot override `bar` because it already has a `final` definition in the trait
+}
+
+fn main() {}

--- a/tests/ui/traits/final/overriding-rpitit.stderr
+++ b/tests/ui/traits/final/overriding-rpitit.stderr
@@ -1,0 +1,14 @@
+error: cannot override `bar` because it already has a `final` definition in the trait
+  --> $DIR/overriding-rpitit.rs:14:5
+   |
+LL |     fn bar() -> impl Clone {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: `bar` is marked final here
+  --> $DIR/overriding-rpitit.rs:6:5
+   |
+LL |     final fn bar() -> impl Clone {
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+

